src/Compilers/CSharp/Portable/Syntax/SyntaxTreeDiagnosticEnumerator.cs

```diff
@@ -11,136 +13,121 @@ namespace Microsoft.CodeAnalysis.CSharp
+                        // green nodes to process on the stack), so we have to roll back.  For nodes, we have yet to see the
+                        // leading trivia, so those don't need an adjustment.
+                        int leadingWidthAlreadyCounted = node.IsToken ? node.GetLeadingTriviaWidth() : 0;
 
-                    // don't produce locations outside of tree span
-                    Debug.Assert(_syntaxTree is object);
-                    var length = _syntaxTree.GetRoot().FullSpan.Length;
-                    var spanStart = Math.Min(_position - leadingWidthAlreadyCounted + sdi.Offset, length);
-                    var spanWidth = Math.Min(spanStart + sdi.Width, length) - spanStart;
+                        // don't produce locations outside of tree span
+                        var length = syntaxTree.GetRoot().FullSpan.Length;
+                        var spanStart = Math.Min(position - leadingWidthAlreadyCounted + sdi.Offset, length);
+                        var spanWidth = Math.Min(spanStart + sdi.Width, length) - spanStart;
 
-                    _current = new CSDiagnostic(sdi, new SourceLocation(_syntaxTree, new TextSpan(spanStart, spanWidth)));
+                        yield return new CSDiagnostic(sdi, new SourceLocation(syntaxTree, new TextSpan(spanStart, spanWidth)));
+                    }
 
-                    _stack.UpdateDiagnosticIndexForStackTop(diagIndex);
-                    return true;
+                    stack.MarkProcessedDiagnosticsForStackTop();
                 }
 
-                var slotIndex = _stack.Top.SlotIndex;
-tryAgain:
-                if (slotIndex < node.SlotCount - 1)
-                {
-                    slotIndex++;
-                    var child = node.GetSlot(slotIndex);
-                    if (child == null)
-                    {
-                        goto tryAgain;
-                    }
+                if (tryContinueWithThisNode(node))
+                    continue;
 
-                    if (!child.ContainsDiagnostics)
-                    {
-                        _position += child.FullWidth;
-                        goto tryAgain;
-                    }
+                // Weren't able to continue with this node. Pop it so we continue processing its parent.
+                stack.Pop();
+            }
+
+            yield break;
 
-                    _stack.UpdateSlotIndexForStackTop(slotIndex);
-                    _stack.PushNodeOrToken(child);
+            bool tryContinueWithThisNode(GreenNode node)
+            {
+                // SlotCount is 0 when we hit tokens or normal trivia. In this case, we just want to move past the
+                // normal width of the item.  Importantly, in the case of tokens, we don't want to move the
+                // full-width.  That would make us double-count the widths of the leading/trailing trivia which
+                // we're walking into as normal green nodes.
```

Comment:
added comments to make the logic/intent clearer (and teh code is still shorter) :)